libraries/retrying_amqp_connection/include/eosio/retrying_amqp_connection/retrying_amqp_connection.hpp

```diff
@@ -0,0 +1,268 @@
+      start_connection();
+   }
+
+   const AMQP::Address& address() const {
+      return _address;
+   }
+
+protected:
+   void onReady(AMQP::Connection* connection) override {
+      fc_ilog(_logger, "AMQP connection to ${s} is fully operational", ("s", _address));
+
+      _ready_callback(connection);
+      _indicated_ready = true;
+   }
+
+   void onData(AMQP::Connection* connection, const char* data, size_t size) override {
+      if(!_sock.is_open())
+         return;
+      _state->outgoing_queue.emplace_back(data, data+size);
+      send_some();
+   }
+
+   void onError(AMQP::Connection* connection, const char* message) override {
+      fc_elog(_logger, "AMQP connection to ${s} suffered an error; will retry shortly: ${m}", ("s", _address)("m", message));
+      schedule_retry();
+   }
+
+   void onClosed(AMQP::Connection *connection) override {
+      fc_wlog(_logger, "AMQP connection to ${s} closed AMQP connection", ("s", _address));
+      schedule_retry();
+   }
+
+private:
+  void start_connection() {
+       _resolver.async_resolve(_address.hostname(), std::to_string(_address.port()), boost::asio::bind_executor(_executor, [this](const auto ec, const auto endpoints) {
+         if(ec) {
+            if(ec != boost::asio::error::operation_aborted) {
+               fc_wlog(_logger, "Failed resolving AMQP server ${s}; will retry shortly: ${m}", ("s", _address)("m", ec.message()));
+               schedule_retry();
+            }
+            return;
+         }
+         //AMQP::Connection's dtor will attempt to send a last gasp message. Resetting state here is a little easier to prove
+         // as being safe as it requires pumping the event loop once vs placing the state reset directly in schedule_retry()
+         _state.emplace();
+         boost::asio::async_connect(_sock, endpoints, boost::asio::bind_executor(_executor, [this](const auto ec, const auto endpoint) {
+            if(ec) {
+               if(ec != boost::asio::error::operation_aborted) {
+                  fc_wlog(_logger, "Failed connecting AMQP server ${s}; will retry shortly: ${m}", ("s", _address)("m", ec.message()));
+                  schedule_retry();
+               }
+               return;
+            }
+            fc_ilog(_logger, "TCP connection to AMQP server at ${s} is up", ("s", _address));
+            receive_some();
+            _state->amqp_connection.emplace(this, _address.login(), _address.vhost());
+         }));
+      }));
+   }
+
+   void schedule_retry() {
+      if(_indicated_ready)
+         _failed_callback();
+      _indicated_ready = false;
+      _sock.close();
+      _resolver.cancel();
+
+      boost::system::error_code ec;
+      _timer.expires_from_now(std::chrono::seconds(1), ec);
+      if(ec)
+         return;
+      _timer.async_wait(boost::asio::bind_executor(_executor, [this](const auto ec) {
+         if(ec)
+            return;
+         start_connection();
```

Comment:
I tested this some to understand the behavior better. Empirically what I'm seeing with an `async_read` & `async_write` both outstanding at the time of a connection failure is both callbacks being called back-to-back with an error code. That does indeed cause two calls in to `schedule_retry()`. Nominally those callbacks are going to come in within 1 second of each other (I don't even see event loop pumped between those callbacks) which will cause the second callback to cancel the existing retry timer before expiry, leaving just a single call to `start_connection()` on the new expiry.

Though, if the timing is _really_ ugly, I suppose the timer started by the first callback could expire before the second backback resets the timer. Which would stack up two `start_connection()`s because the first expiry callback is already queued and can not be cancelled.